File: crates/clang-importer-sys/cxx/lib/cxx/llvm/Twine.cxx
#include "cxx/llvm/Twine.hxx"

namespace cxx {
namespace llvm {
namespace Twine {

std::shared_ptr<::llvm::Twine> make() {
  return std::make_shared<::llvm::Twine>(::llvm::Twine::createNull());
}

std::shared_ptr<::llvm::Twine> from_cxx_string(std::string const &Str) {
  return std::make_shared<::llvm::Twine>(::llvm::Twine(Str));
}

std::shared_ptr<::llvm::Twine>
from_string_ref(rust::llvm::StringRef const &Str) {
  return std::make_shared<::llvm::Twine>(::llvm::Twine(*Str.ptr));
}

} // namespace Twine
} // namespace llvm
} // namespace cxx

File: crates/clang-importer-sys/tests/llvm/twine.rs
use clang_importer_sys::llvm::{StringRef, Twine};
use cxx::let_cxx_string;
use tap::prelude::*;

#[test]
fn new() {
    unsafe {
        let _ = Twine::new();
    }
}

#[test]
fn from_cxx_string() {
    unsafe {
        let_cxx_string!(expected = "string");
        assert_eq!(&*expected, &*Twine::from(&*expected).str());
    }
}

#[test]
fn from_string_ref() {
    unsafe {
        let value = "string";
        let_cxx_string!(expected = value);
        assert_eq!(&*expected, &*Twine::from(StringRef::from(value)).str());
    }
}
